### ISSUE TITLE
PKCE code verifier doesn't meet rfc7636 spec

### ISSUE THREAD

**Issue by schalk-b** [AzureAD/microsoft-authentication-library-for-js#1501]
## Library
- [ ] `msal@1.2.1` or `@azure/msal@1.2.1`
- [x] `@azure/msal-browser@2.x.x`
- [ ] `@azure/msal-angular@0.x.x`
- [ ] `@azure/msal-angular@1.x.x`
- [ ] `@azure/msal-angularjs@1.x.x`

## Description
Shouldn't the PKCE code verifier meet the rfc7636 spec?

Specifically in the spec it mentions

> code_verifier = high-entropy cryptographic random STRING using the
>    unreserved characters [A-Z] / [a-z] / [0-9] / "-" / "." / "_" / "~"
>    from Section 2.3 of [RFC3986], with a minimum length of 43 characters
>    and a maximum length of 128 characters.
> 

However, reviewing the code #1141 for the PkceGenerator I noticed a few things.

1. The code verifier is always 32 characters long (instead of the required minimum of 43).
2. The code verifier characters aren't uniformly random. 
```
/**
  * Generates a character string based on input array.
  * @param buffer 
  */
 private bufferToCVString(buffer: Uint8Array): string {
     const charArr = [];
     for (let i = 0; i < buffer.byteLength; i += 1) {
         const index = buffer[i] % CV_CHARSET.length;
         charArr.push(CV_CHARSET[index]);
     }
     return charArr.join("");
 }
```

Because we are using modulus `buffer[i] % CV_CHARSET.length` it means from our character set, all characters before '6' have a 4/256 chance of showing, '6' and after, have a 3/256 change of showing.

I know that the current implementation doesn't pose a 'serious' security risk. But I think it's worth pointing out that there is room for improvement.

**Comment by schalk-b** [comment]
I would be happy to take this one on and submit a PR if you want me to?

It's just a matter of implementing a binary encoder for the character set in the spec: [A-Z] / [a-z] / [0-9] / "-" / "." / "_" / "~"

**Comment by pkanher617** [comment]
@schalk-b Sorry for missing this PR. I would be happy to review a PR that you have submitted. Please give me some time today to review the PKCE generator code and I can get back to you with a more informed answer to your original question.

**Comment by pkanher617** [comment]
@schalk-b Thanks again for opening this issue.

As far as I can see, the lengths of the challenge and verifier that get generated are 43 characters. I used tests in the browser as well as reviewed the network traces in different browsers.

```javascript
it.only("generatePkceCode()", async () => {
        sinon.stub(BrowserCrypto.prototype, <any>"getSubtleCryptoDigest").callsFake(async (algorithm: string, data: Uint8Array): Promise<ArrayBuffer> => {
            expect(algorithm).to.be.eq("SHA-256");
            return crypto.createHash("SHA256").update(Buffer.from(data)).digest();
        });

        /**
         * Contains alphanumeric, dash '-', underscore '_', plus '+', or slash '/' with length of 43.
         */
        const regExp = new RegExp("[A-Za-z0-9-_+/]{43}");
        const generatedCodes: PkceCodes = await cryptoObj.generatePkceCodes();
        console.log(generatedCodes);
        console.log(generatedCodes.challenge.length);
        console.log(generatedCodes.verifier.length);
        expect(regExp.test(generatedCodes.challenge)).to.be.true;
        expect(regExp.test(generatedCodes.verifier)).to.be.true;
    });
```

![codeChallengeScreenshot](https://user-images.githubusercontent.com/6913227/79488541-baab4b00-7fce-11ea-9598-a35fa0d242ce.png)

![codeverifierScreenshot](https://user-images.githubusercontent.com/6913227/79488637-ea5a5300-7fce-11ea-8852-591e80b6eee4.png)

Could you let me know what browser and environment you are using?

The randomness is something I am open to reviewing further. Could you explain how you arrived at the values for 4/256 and 3/256?

**Comment by schalk-b** [comment]
@pkanher617 Thanks for spending your time on this issue.

You are right that when calling generatePkceCodes(); the verifier is actually 43, as I only just noticed  it encodes the verifier in base64. However, it does the base64 encoding, after running the random octet sequence through bufferToCVString.

To this is currently what happens:
1. Random 32 octet sequence
    buffer = [21, 231, ..., 232]
2. bufferToCVString(buffer)
    verifyString = uzxqicc89jAuURxH4qtzsfJH0YS2Jy.t
3. base64urlEncode(verifyString)
    dXp4cWljYzg5akF1VVJ4SDRxdHpzZkpIMFlTMkp5LnQ

I believe that the bufferToCVString is unnecessary and basically takes some of the random, out of the random buffer (I'll show a test to demonstrate this at the end of the comment).

We can change:
```
/**
     * Generates a random 32 byte buffer and returns the base64
     * encoded string to be used as a PKCE Code Verifier
     */
    private generateCodeVerifier(): string {
        try {
            // Generate random values as utf-8
            const buffer: Uint8Array = new Uint8Array(RANDOM_BYTE_ARR_LENGTH);
            this.cryptoObj.getRandomValues(buffer);
            // verifier as string
            const pkceCodeVerifierString = this.bufferToCVString(buffer);
            // encode verifier as base64
            const pkceCodeVerifierB64: string = this.base64Encode.urlEncode(pkceCodeVerifierString);
            return pkceCodeVerifierB64;
        } catch (e) {
            throw BrowserAuthError.createPkceNotGeneratedError(e);
        }
    }
```

to

```
/**
     * Generates a random 32 byte buffer and returns the base64
     * encoded string to be used as a PKCE Code Verifier
     */
    private generateCodeVerifier(): string {
        try {
            // Generate random values as utf-8
            const buffer: Uint8Array = new Uint8Array(RANDOM_BYTE_ARR_LENGTH);
            this.cryptoObj.getRandomValues(buffer);
            // encode verifier as base64
            const pkceCodeVerifierB64: string = this.base64Encode.urlEncodeArr(pkceCodeVerifierString);
            return pkceCodeVerifierB64;
        } catch (e) {
            throw BrowserAuthError.createPkceNotGeneratedError(e);
        }
    }
```

And then we can completely remove the bufferToCVString method.

**Comment by schalk-b** [comment]
Sorry for this very long an ugly test, but this shows that the current bufferToCVString method that gets called removes some of the random, by not encoding correctly.

```javascript
import { expect } from "chai";
import { BrowserCrypto } from "../../src/crypto/BrowserCrypto";
import { PkceGenerator } from "../../src/crypto/PkceGenerator";

describe("PkceVerifier.spec.ts Unit Tests", () => {
    it("bufferToCVString() evenly distributed encoding", async () => {
      // This only works with a very large number of tests as we are testing distribution from random numbers
      const NUM_TESTS = 100000;

      // Added here because they aren't exported from PkceGenerator
      const CV_CHARSET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~";
      const RANDOM_BYTE_ARR_LENGTH = 32;

      /** Arrange */
      const browserCrypto = new BrowserCrypto();
      const pkceGenerator = new PkceGenerator(browserCrypto);

      // We want to count the amount of times each character is encoded.
      const characterCounts = {};
      [...CV_CHARSET].forEach(char => characterCounts[char] = 0);
      let totalCharacters = 0;

      // Asserting with randomly generated numbers are hard, so lets add a margin of error
      // but basically we expect the change of each character to be encoded to be 1/CV_CHARSET.length
      const marginOfError = 0.0001;
      const min = 1 / CV_CHARSET.length - marginOfError;
      const max = 1 / CV_CHARSET.length + marginOfError;

      // Act
      for (let i = 0; i < 100000; i++) {
        const buffer: Uint8Array = new Uint8Array(RANDOM_BYTE_ARR_LENGTH);
        browserCrypto.getRandomValues(buffer);

        // verifier as string (not public, so we have to use ['bufferToCVString'])
        // COMMENT TO MAKE TEST PASS
        const pkceCodeVerifierString = pkceGenerator['bufferToCVString'](buffer);

        // UNCOMMENT TO MAKE TEST PASS
        // const pkceCodeVerifierString = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~";

        // counting
        totalCharacters += pkceCodeVerifierString.length;
        [...pkceCodeVerifierString].forEach(char => characterCounts[char]++);
      }

      // Logging
      console.log(`Expected range: ${min.toFixed(4)}-${max.toFixed(4)}`);
      [...CV_CHARSET].forEach(char => {
        const distribution = Math.round(characterCounts[char] / totalCharacters * 10000) / 10000;
        console.log(`${char}: ${distribution}`);
      });
      
      // Assert
      [...CV_CHARSET].forEach(char => {
        const distribution = Math.round(characterCounts[char] / totalCharacters * 10000) / 10000;
        expect(distribution).to.be.gte(min);
        expect(distribution).to.be.lte(max);
      });
    });
});
```

**Comment by schalk-b** [comment]
You will see the distribution all the way through to 5 is about 0.0156 and then 6 onward is 0.0117.

**Comment by schalk-b** [comment]
Are you open to accepting a PR for this one? Or does it need further review?

**Comment by pkanher617** [comment]
Merged in #1872
